pkg/runtime/pubsub/default_bulkpub.go

```diff
@@ -0,0 +1,132 @@
+	"github.com/dapr/dapr/utils"
+)
+
+const (
+	bulkPublishSeriallyKey       string = "bulkPublishSerially"
+	bulkPublishMaxConcurrencyKey string = "bulkPublishMaxConcurrency"
+
+	defaultBulkPublishMaxConcurrency int = 100
+)
+
+var ErrBulkPublishFailure = errors.New("bulk publish failed")
+
+// defaultBulkPublisher is the default implementation of BulkPublisher.
+// It is used when the component does not implement BulkPublisher.
+type defaultBulkPublisher struct {
+	p contribPubsub.PubSub
+}
+
+// NewDefaultBulkPublisher returns a new defaultBulkPublisher from a PubSub.
+func NewDefaultBulkPublisher(p contribPubsub.PubSub) *defaultBulkPublisher {
+	return &defaultBulkPublisher{
+		p: p,
+	}
+}
+
+// BulkPublish publishes a list of messages to a topic as individual Publish requests.
+// If 'bulkPublishSerially' metadata is set to true, the messages are sent to the broker serially,
+// in the same order. Otherwise they are sent to the broker as parallel Publish requests.
+func (p *defaultBulkPublisher) BulkPublish(_ context.Context, req *contribPubsub.BulkPublishRequest) (contribPubsub.BulkPublishResponse, error) {
+	if req.Metadata[bulkPublishSeriallyKey] == "true" {
+		return p.bulkPublishSerial(req)
+	} else {
+		return p.bulkPublishParallel(req)
+	}
+}
+
+// bulkPublishSerial publishes messages in a serial order. This is slower, but ensures
+// that messages are sent to the broker in the same order as specified in the request.
+func (p *defaultBulkPublisher) bulkPublishSerial(req *contribPubsub.BulkPublishRequest) (contribPubsub.BulkPublishResponse, error) {
+	statuses := make([]contribPubsub.BulkPublishResponseEntry, 0, len(req.Entries))
+	hasError := false
+
+	for _, entry := range req.Entries {
+		status := p.bulkPublishSingleEntry(req.PubsubName, req.Topic, entry)
+		if status.Error != nil {
+			hasError = true
+		}
+
+		statuses = append(statuses, status)
+	}
+
+	var err error
+	if hasError {
+		err = ErrBulkPublishFailure
+	}
+
+	return contribPubsub.BulkPublishResponse{Statuses: statuses}, err
+}
+
+// bulkPublishParallel publishes messages in parallel. This is faster, but does not guarantee
+// that messages are sent to the broker in the same order as specified in the request.
+func (p *defaultBulkPublisher) bulkPublishParallel(req *contribPubsub.BulkPublishRequest) (contribPubsub.BulkPublishResponse, error) {
+	statuses := make([]contribPubsub.BulkPublishResponseEntry, 0, len(req.Entries))
+	var eg errgroup.Group
+	eg.SetLimit(utils.GetIntOrDefault(req.Metadata, bulkPublishMaxConcurrencyKey, defaultBulkPublishMaxConcurrency))
+
+	statusChan := make(chan contribPubsub.BulkPublishResponseEntry, len(req.Entries))
+
+	for _, entry := range req.Entries {
+		entry := entry
+		eg.Go(func() error {
+			status := p.bulkPublishSingleEntry(req.PubsubName, req.Topic, entry)
+			statusChan <- status
+			return status.Error
+		})
+	}
+
+	err := eg.Wait()
+	close(statusChan)
+
+	for status := range statusChan {
```

Comment:
If you move this before `err := eg.Wait()` you should be able to use an unbuffered channel (`statusChan := make(chan contribPubsub.BulkPublishResponseEntry)`) - which at the very least uses less memory.

Reply:
Actually looking at this now, you are closing `statusChan` before reading it (line above is `close(statusChan)`) so this range will always read from a closed channel. You do need to move this up and use an unbuffered chan.

Reply:
With unbuffered channels, if `eg` has a limit of 100 and N is 1000, it will spawn 100 goroutines and get blocked at `eg.Go(fn)`.

Updated the `statusChan`'s buffer size to `eg`'s limit to save memory, and moved `range` before `eg.Wait()` to read in parallel. Thoughts? @ItalyPaleAle 

Reading from a closed channel should be fine, right? Channels should be closed whenever the sender is done sending. `range` can read all items safely. https://gobyexample.com/range-over-channels

Reply:
Right, reading from a closed channel is ok if the channel is buffered.

I see you're still using buffered channels but less capacity. Should be ok, although:

> With unbuffered channels, if eg has a limit of 100 and N is 1000, it will spawn 100 goroutines and get blocked at eg.Go(fn).

AFAIK eg's concurrency is limited by the max concurrency you configure it with, not by the capacity of the channel where the results are stored. An unbuffered channel should still work because the "main" goroutine reads from it (`range statusChan`) and other goroutines can't write in the channel until it's read.

But this is fine anyways.

Reply:
Moving this conversation to #5396.